lib/compilers/ewavr.js

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2018, Ethan Slattery
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+"use strict";
+
+const BaseCompiler = require('../base-compiler'),
+    temp = require('temp'),
+    path = require('path'),
+    _ = require('underscore'),
+    PELabelReconstructor = require("../pe32-support").labelReconstructor,
+    AsmParser = require('../asm-parser-ewavr.js');
+
+class EWAVRCompiler extends BaseCompiler {
+    constructor(info, env) {
+        super(info, env);
+        this.asm = new AsmParser(this.compilerProps);
+    }
+
+    newTempDir() {
```

Comment:
not really directly related to your PR, but it seems silly to have to override and reimplement this function just to change `TMPDIR` to `TMP` because this is expected to run on windows when `os.tmpdir()` is part of the node API and `temp.mkdir` defaults to using `os.tmpdir()` when a directory is not specified.

As far as I can tell this could be refactored to
```js
temp.mkdir('compiler-explorer-compiler', (err, dirPath) => { ... });
```
in BaseCompiler and the meaning would be identical and remove the need to override for win32 compilers.

Maybe we should refactor anywhere else `process.env.TMP` (or similar) is being used to `os.tmpdir()` - you can see the implementation in node basically just reads env variables for the appropriate os with a hardcoded fallback.

https://github.com/nodejs/node/blob/f2060fa9e6d7b179c3c49b2e630e3bc981a565de/lib/os.js#L126

Reply:
it's been a while, but yes. I had to copy this from MSVC if I remember correctly.